### PR TITLE
Contract Endpoint

### PR DESCRIPTION
**Pull request by KMayne**
# What?
The API endpoint /contracts/:contract. Returns an object representing a contract - for now it only holds the variables.

### CODE REVIEW

File: apps/etheroscope/lib/etheroscope/contract/contract.ex [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  def abi_item_is_variable(abi_item) do
+    abi_item["type"] == "function" and length(abi_item["inputs"]) == 0
+    and length(abi_item["outputs"]) == 1
+    and Enum.member?(@allowed_types, Enum.at(abi_item["outputs"], 0)["type"])
+  end
+
+  def fetch_contract(contract_address) do
+    IO.inspect Application.get_env(:etheroscope, :etherscan_api_key)
+    url = "https://api.etherscan.io/api?module=contract&action=getabi&address="
```
Comment (Thopiax) [inline apps/etheroscope/lib/etheroscope/contract/contract.ex]: `url = "https://api.etherscan.io/api?module=contract&action=getabi&address=#{contract_address}&apikey=#{Application.get_env(:etheroscope, :etherscan_api_key)}"` might be better. Both work though 😄

File: apps/etheroscope/lib/etheroscope/contract/contract.ex [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      <> contract_address <> "&apikey=" <> Application.get_env(:etheroscope, :etherscan_api_key)
+    {:ok, response} = HTTPoison.get(url)
```
Comment (Thopiax) [inline apps/etheroscope/lib/etheroscope/contract/contract.ex]: Either `response = HTTPoison.get!(url)` which will raise an error if it fails or 
`case HTTPoison.get(url) do
    {:ok, response} -> # ....
    {:error, resp}     -> handle error
end`

File: apps/etheroscope/lib/etheroscope/contract/contract.ex [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    response_body = Poison.decode!(response.body)
+
+    if (response_body["result"] == "") do
+      %{:variables => [], :error => "No ABI found"}
+    else
+      contract_abi = Poison.decode!(response_body["result"])
+      variables = Enum.filter(contract_abi, &abi_item_is_variable/1)
+      var_names = Enum.map(variables, fn (var) -> var["name"] end);
+      %{:variables => var_names}
```
Comment (Thopiax) [inline apps/etheroscope/lib/etheroscope/contract/contract.ex]: I would probably return `{:ok, %{}}` or `{:error, %{}}` to make pattern matching easier for the caller.

File: README.md [hunk 4]
```diff
@@ -9,6 +9,19 @@ In order to get your environment setup for development, follow these instruction
   2. Build a docker image with the `Dockerfile` and tag it `etheroscope` (try `docker build . -f Dockerfile -t etheroscope`)
   3. Run `docker-compose up -d`.
 
+### Running without Docker
```
Comment (Thopiax) [inline README.md]: 👍

**Review by Thopiax — CHANGES_REQUESTED** [review]

File: apps/etheroscope/lib/etheroscope/contract/contract.ex (hunk 1 above)
Comment (KMayne) [reply apps/etheroscope/lib/etheroscope/contract/contract.ex]: Ah I didn't know Elixir had template strings 👍

File: apps/etheroscope/lib/etheroscope/contract/contract.ex (hunk 3 above)
Comment (KMayne) [reply apps/etheroscope/lib/etheroscope/contract/contract.ex]: Fair enough

File: apps/etheroscope/lib/etheroscope/contract/contract.ex (hunk 2 above)
Comment (KMayne) [reply apps/etheroscope/lib/etheroscope/contract/contract.ex]: Yes will add this

**Review by Thopiax — APPROVED** [review]

**Pull request merged**